src/vdaf/poplar1.rs

```diff
@@ -128,6 +130,24 @@ pub struct Poplar1InputShare<const SEED_SIZE: usize> {
     corr_leaf: [Field255; 2],
 }
 
+impl<const SEED_SIZE: usize> ConstantTimeEq for Poplar1InputShare<SEED_SIZE> {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        // We short-circuit on the length of corr_inner being different. Only the content is
+        // protected.
+        if self.corr_inner.len() != other.corr_inner.len() {
+            return Choice::from(0);
+        }
+
+        let mut rslt = self.idpf_key.ct_eq(&other.idpf_key)
+            & self.corr_seed.ct_eq(&other.corr_seed)
+            & self.corr_leaf.ct_eq(&other.corr_leaf);
+        for (x, y) in self.corr_inner.iter().zip(other.corr_inner.iter()) {
```

Comment:
We can avoid manually checking lengths and iterating by using the blanket [`impl<T: ConstantTimeEq> ConstantTimeEq for [T]`](https://docs.rs/subtle/latest/subtle/trait.ConstantTimeEq.html#impl-ConstantTimeEq-for-%5BT%5D). Just as the code here does, it short-circuits on slice length. Then, this method body could be simplified to a single four-way `&` expression.

Reply:
I tried this, unfortunately it doesn't work with an error like:

```
error[E0599]: the method `ct_eq` exists for struct `Vec<[Field64; 2]>`, but its trait bounds were not satisfied
   --> src/vdaf/poplar1.rs:159:31
    |
159 |             & self.corr_inner.ct_eq(&other.corr_inner)
    |                               ^^^^^ method cannot be called on `Vec<[Field64; 2]>` due to unsatisfied trait bounds
    |
    = note: the following trait bounds were not satisfied:
            `[Field64; 2]: ConstantTimeEq`
            which is required by `[[Field64; 2]]: ConstantTimeEq`
```

I _think_ the reason `[Field64; 2]` doesn't satisfy `ConstantTimeEq` is that it can't be automatically converted to a `[Field64]`. `subtle`'s `const-generics` feature seems like it should help -- but it actually only provides a blanket implementation of `ConditionallySelectable` for `[T; N]`, no `ConstantTimeEq` implementation. I don't see why one could be written, but this would require waiting on a release of the `subtle` crate.

Reply:
I filed https://github.com/dalek-cryptography/subtle/pull/114, we'll see if it lands. The last release was 6 months ago so we may be waiting awhile in any case.